nse_libssh2.cc

```diff
@@ -0,0 +1,949 @@
+    }
+    if (rc != 0)
+        return luaL_error(L, "Requesting pty");
+
+    return 1;
+}
+
+static int l_setup_channel (lua_State *L) {
+    return setup_channel(L, 0, 0);
+}
+
+static int finish_open_channel (lua_State *L, int status, lua_KContext ctx) {
+    ssh_userdata *state = (ssh_userdata *)lua_touserdata(L, 1);
+    LIBSSH2_CHANNEL **channel = (LIBSSH2_CHANNEL **) lua_touserdata(L, 2);
+
+    while ((*channel = libssh2_channel_open_session(state->session)) == NULL
+    && libssh2_session_last_errno(state->session) == LIBSSH2_ERROR_EAGAIN) {
+        luaL_getmetafield(L, 1, "filter");
+        lua_pushvalue(L, 1);
+        lua_callk(L, 1, 0, 0, finish_open_channel);
+    }
+    if (channel == NULL)
+        return luaL_error(L, "Opening channel");
+
+    return setup_channel(L, 0, 0);
+}
+
+static int l_open_channel (lua_State *L) {
+    ssh_userdata *state = (ssh_userdata *)lua_touserdata(L, 1);
+    LIBSSH2_CHANNEL **channel = (LIBSSH2_CHANNEL **)lua_newuserdata(L, sizeof(LIBSSH2_CHANNEL *));
+
+    while ((*channel = libssh2_channel_open_session(state->session)) == NULL
+    && libssh2_session_last_errno(state->session) == LIBSSH2_ERROR_EAGAIN) {
+        luaL_getmetafield(L, 1, "filter");
+        lua_pushvalue(L, 1);
+        lua_callk(L, 1, 0, 0, finish_open_channel);
+    }
+
+    return l_setup_channel(L);
+}
+
+static int channel_close (lua_State *L, int status, lua_KContext ctx) {
+    int rc;
+    // ssh_userdata *state = (ssh_userdata *)lua_touserdata(L, 1);
+    LIBSSH2_CHANNEL **channel = (LIBSSH2_CHANNEL **) lua_touserdata(L, 2);
+
+    while ((rc = libssh2_channel_close(*channel)) == LIBSSH2_ERROR_EAGAIN) {
+        luaL_getmetafield(L, 1, "filter");
+        lua_pushvalue(L, 1);
+        lua_callk(L, 1, 0, 0, channel_close);
+    }
+    if (rc != 0)
+        return luaL_error(L, "Error closing channel");;
+
+    return 0;
+}
+
+static int l_channel_close (lua_State *L) {
+    return channel_close(L, 0, 0);
+}
+
+static const struct luaL_Reg libssh2[] = {
+    { "session_open", l_session_open },
+    { "hostkey_hash", l_hostkey_hash },
+    { "set_timeout", l_set_timeout },
+    { "userauth_list", l_userauth_list },
+    { "userauth_publickey", l_userauth_publickey },
+    { "read_publickey", l_read_publickey },
+    { "publickey_canauth", l_publickey_canauth },
+    { "userauth_password", l_userauth_password },
+    { "session_close", l_session_close },
+    { "open_channel", l_open_channel},
+    { "channel_read", l_channel_read},
+    { "channel_read_stderr", l_channel_read_stderr},
+    { "channel_write", l_channel_write},
+    { "channel_exec", l_channel_exec},
+    { "channel_send_eof", l_channel_send_eof},
+    { "channel_eof", l_channel_eof},
+    { "channel_close", l_channel_close},
+    { NULL, NULL }
+};
+
+static int gc (lua_State *L) {
+    struct ssh_userdata *sshu = NULL;
+
+    sshu = (struct ssh_userdata *) nseU_checkudata(L, 1, SSH2_UDATA, "ssh2");
+    if (sshu) {
```

Comment:
This condition should also cover the closing of the sockets below, otherwise if `sshu` is `NULL` then it's a null pointer dereference. This line could be written as `if (!sshu) { return 0; }` to avoid indenting the rest of the function, or even `return luaL_error(L, "some error");`, though I am not sure what the exact sequence of events would have to be in order for it to be `NULL`.